clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -264,44 +264,81 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
   return DeviceLibFiles;
 }
 
-/// Link all device library files and input file into one LLVM IR file. This
-/// linking is performed using llvm-link tool.
-/// 'InputFiles' is the list of all LLVM IR device input files.
-/// 'Args' encompasses all arguments required for linking device code and will
-/// be parsed to generate options required to be passed into llvm-link tool.
-static Expected<StringRef> linkDeviceLibFiles(StringRef InputFile,
-                                              const ArgList &Args) {
-  llvm::TimeTraceScope TimeScope("LinkDeviceLibraryFiles");
+/// Following tasks are performed:
+/// 1. Link all SYCL device bitcode images into one image. Device linking is
+/// performed using the linkInModule API.
+/// 2. Gather all SYCL device library bitcode images.
+/// 3. Link all the images gathered in Step 2 with the output of Step 1 using
+/// linkInModule API. LinkOnlyNeeded flag is used.
+Expected<StringRef> linkDeviceCode(ArrayRef<std::string> InputFiles,
+                                   const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("SYCL link device code");
+  LLVMContext C;
+
+  assert(InputFiles.size() && "No inputs to link");
+
+  // Create a new file to write the linked device file to
+  auto BitcodeOutput =
+      createTempFile(Args, sys::path::filename(OutputFile), "bc");
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
 
+  // Get all SYCL device library files, if any
   auto SYCLDeviceLibFiles = getSYCLDeviceLibs(Args);
   if (!SYCLDeviceLibFiles)
     return SYCLDeviceLibFiles.takeError();
-  if ((*SYCLDeviceLibFiles).empty())
-    return InputFile;
 
-  Expected<std::string> LLVMLinkPath =
-      findProgram(Args, "llvm-link", {getMainExecutable("llvm-link")});
-  if (!LLVMLinkPath)
-    return LLVMLinkPath.takeError();
+  if (DryRun || Verbose) {
+    std::string Inputs =
+        std::accumulate(std::next(InputFiles.begin()), InputFiles.end(),
+                        InputFiles.front(), [](std::string a, std::string b) {
+                          return std::move(a) + ", " + std::move(b);
+                        });
+    std::string LibInputs = "";
+    if (!(*SYCLDeviceLibFiles).empty())
+      LibInputs = std::accumulate(
+          std::next((*SYCLDeviceLibFiles).begin()), (*SYCLDeviceLibFiles).end(),
+          (*SYCLDeviceLibFiles).front(), [](std::string a, std::string b) {
+            return std::move(a) + ", " + std::move(b);
+          });
+    errs() << formatv(
+        "sycl-device-link: inputs: {0} libfiles: {1} output: {2}\n", Inputs,
+        LibInputs, *BitcodeOutput);
+    if (DryRun)
+      return *BitcodeOutput;
+  }
```

Comment:
According to my understanding, we should bailout in dry-run anymore as don't call the external tool. We should execute the function in any mode.
We probably should have some option to output the result of the linking to validate it.

Reply:
Added -print-linked-module option and also a test. Thanks